adam/visualization/positioning.py

```diff
@@ -0,0 +1,371 @@
+        max_indices = torch.max(combined_mins, 1)
+        maximum_mins = torch.take(combined_mins, max_indices[1] + (dims * 2))
+
+        # should stick together the minimum parts and the maximum parts
+        # with columns like:
+        # [ min0x   min1x
+        #   min0y   min1y
+        #   min0z   min1z
+        #                ]
+        # then find the maximum element from each row
+
+        # repeat the process for the min of the max projections
+        maxs_0 = min_max_proj_0.gather(1, torch.tensor([[1], [1], [1]]))
+        maxs_1 = min_max_proj_1.gather(1, torch.tensor([[1], [1], [1]]))
+        combined_maxes = torch.stack((maxs_0, maxs_1), 1).squeeze()
+        min_indices = torch.min(combined_maxes, 1)
+        minimum_maxes = torch.take(combined_maxes, min_indices[1] + (dims * 2))
+
+        return torch.stack((maximum_mins, minimum_maxes), 1)
+
+    @staticmethod
+    def overlap_penalty(min_max_overlaps: torch.Tensor) -> torch.Tensor:
+        """
+
+        Args:
+            min_max_overlaps: (3, 2) tensor -> intervals of how much the boxes overlap in each dimension
+
+        Returns: Tensor with a scalar of the collision penalty size
+
+        """
+        # subtract each minimum max from each maximum min:
+        overlap_distance = min_max_overlaps[:, 0] - min_max_overlaps[:, 1]
+
+        # if ANY element in overlap_distance is positive, the minimum positive value
+        # then the two are not colliding
+
+        # otherwise the penetration distance is the maximum negative value
+        # (the smallest translation that would disentangle the two
+
+        # if not colliding
+        for dim in range(3):
+            if overlap_distance[dim] >= 0:
+                # multiplication by zero in order to still return a Tensor
+                return overlap_distance[dim] * 0
+
+        smallest_colliding_dim = 0
+        for dim in range(3):
+            if 0 > overlap_distance[dim] >= overlap_distance[smallest_colliding_dim]:
+                smallest_colliding_dim = dim
+
+        return overlap_distance[smallest_colliding_dim] * -10 + 1
+
+
+class PositionSolver:
+    # constants to eventually use
+    POSITION_CONSTRAINTS = {
+        "X_RANGE": (-10.0, 10.0),
+        "Y_RANGE": (-10.0, 5.0),
+        "Z_RANGE_GROUND": (0.0, 0.0),
+        "Z_RANGE_FLY": (2.0, 5.0),
+        "Z_RANGE_OTHER": (0.0, 3.0),
+    }
+
+    CLOSE_DISTANCE = 0.25
+
+    def __init__(self, main_point: ndarray, points: List[ndarray]) -> None:
+        self.main_position = main_point
+        self.positions = points
+
+
+class CollisionPenalty(nn.Module):
+    def __init__(self):
+        super().__init__()
+
+    def forward(
+        self,
+        bounding_box_1: AxisAlignedBoundingBox,
+        bounding_box_2: AxisAlignedBoundingBox,
+    ) -> torch.Tensor:
+
+        # get face norms from one of the boxes:
+        face_norms = bounding_box_2.get_face_norms()
+
+        return AxisAlignedBoundingBox.overlap_penalty(
```

Comment:
I would shift all the overlap logic into this class, rather than on `AxisAlignedBoundingBox`

Reply:
Also this is complicated enough to need some explanation (which could be a reference to somewhere else which explains it in detail)